### DISCUSSION TITLE
UserObject with an list of sub-objects

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#27177] [Q&A General] [answered]
Dear Moose community,

I'd like to build an UserObject with an list of nested sub-objects. I hope you can point me in the right direction.

I know how to register register 'simple' parameters for a user object, e.g.:
`  params.addParam<FileName>("data_file", "File holding CSV data");`

This allows me to use this parameter in my input file (*.i - file) like this:
```
  [UserObjects]
    [MyUserObject1]
      type = 'MyUserObject'
      data_file = 'bla.txt'
    []
  []
```

Now I want to build something like this: my user object contains a list of sub-objects beside my 'regular' properties
```
  [UserObjects]
    [MyUserObject1]
      type = 'MyUserObjectType'
      data_file = 'bla.txt'
      [MySubObject1]
        type = 'MySubObjectTypeA'
        propA = 'blubb'
      []
      [MySubObject2]
        type = 'MySubObjectTypeB'
        propY = 888
      []
      [MySubObject2]
        type = 'MySubObjectTypeC'
        propZ = 123.2
      []
      # any more sub-objects as desired by the user...
    []
  []
```

In my c++ object of type 'MyUserObjectType' it would be nice to have all sub-objects in an list/array/vector. 

How can I realize this?

Kind regards,
Jörg

**Comment by GiudGiud** [comment, marked as answer]
Hello

You want to create a new Action. See this page
https://mooseframework.inl.gov/source/actions/Action.html
and this part of the tutorial:
https://mooseframework.inl.gov/workshop/index.html#/57

You probably want to create a new syntax for adding user objects. In this case I would recommend differentiating it from our 
```
[UserObjects]
```
block to avoid confusion and to avoid having the Parser read `[UserObjects/MyUserObject]` and create a regular user object. 

So maybe
```
[NestedUserObjects]
  [parent_uo]
    [child1]
    []
    [child2]
    []
  []
[]
```

**Reply by jmeier** [reply]
Dear @GiudGiud

In the action (more specific in the "act" method) creating "child1" and "child2", how do I get the superordinated "parent_uo"? I know about `_type` and `_name` - but how about the "parent"?

**Reply by GiudGiud** [reply]
Not sure but I would try
- action warehouse to get the name of the current syntax (might have to use the task for that)
- Get the user object (a pointer to) from the problem

**Reply by jmeier** [reply]
Thanks for the help, @GiudGiud. Now I do the following (within the `act` method):
- From the action warehouse I get the current action name (`getCurrentActionName`). For `child1` this string reads `NestedUserObjects/parent_uo/child1`
- One can split this string using `/` as a delimiter.
- Using `_problem->getUserObject<...>("parent_uo")` one can now get the parental user object.
